README.md
# NAME

Log::History - Make self-documenting scripts that track their own execution history.

# SYNOPSIS

Place the import statement for Log::History wherever you want the log to start.

    use Log::History;
    #2014-02-02 12:43:08 (00:00:48) in /path/to/workingdir: /path/to/script.pl --opt 1 --param 2
    ...

To limit the number of log entries, specify how many to keep.
The oldest entries are discarded.

    use Log::History '3';
    #2014-05-27 ...
    #2007-12-18 ...
    #1987-12-18 ...

# DESCRIPTION

Log::History enables scripts to keep track of their own execution history.
Each log entry reports the date, start time, elapsed run time, working directory,
and a record of exactly how the script was called.

# CAVEATS

Because Log::History modifies the script that imports it, problems can occur
if too many instances of the script are run at one time.
(Actually, the problem can arise if one instance ends at the exact time another is starting.)

If you don't have write access for the script, no log will be generated.

# ACKNOWLEDGEMENTS

I was inspired to write this after recently re-reading Neil Bowers' post:
[Identifying CPAN distributions you could help out with](http://blogs.perl.org/users/neilb/2012/12/modules-that-are-candidates-for-helping-out.html).
In it, I found Tushar Murudkar's no-longer-maintained module [Log::SelfHistory](https://metacpan.org/pod/Log::SelfHistory).
I was intrigued since I had been working on [Log::Reproducible](https://github.com/mfcovington/Log-Reproducible).
I wanted to go in a different direction, so decided to start from scratch instead of trying to take over [Log::SelfHistory](https://metacpan.org/pod/Log::SelfHistory).

# AUTHOR

Michael F. Covington <mfcovington@gmail.com>

# SEE ALSO

[Log::SelfHistory](https://metacpan.org/pod/Log::SelfHistory), [Log::Reproducible](https://github.com/mfcovington/Log-Reproducible)


CHANGES
Revision history for Perl module Log::History

0.1.0 2014-06-23

  - Make self-documenting scripts that track their own execution history


Build.pl
#!/usr/bin/env perl
use strict;
use warnings FATAL => 'all';
use Module::Build;

# TODO: Add version specifications for requirements
# TODO: Add perl + version to requirements

my $module_file = "lib/Log/History.pm";

eval "use Pod::Markdown";
if ( ! $@ ) {
    require Pod::Markdown;
    my $readme_file = "README.md";
    pod2markdown( $module_file, $readme_file );
}

my $builder = Module::Build->new(
    module_name        => 'Log::History',
    dist_version_from  => $module_file,
    license            => 'perl',
    create_makefile_pl => 0,
    requires           => {
        "Carp"           => 0,
        "Cwd"            => 0,
        "File::Copy"     => 0,
        "File::Temp"     => 0,
        "POSIX"          => 0,
        "Scalar::Util"      => 0,
    },
    # build_requires => {
    #     'Test::More' => 0,
    # },
    configure_requires => {
        'Module::Build' => 0,
    },
    recommends => {
        # 'Devel::Cover'   => 0,    # To generate testing coverage report
        'Pod::Markdown'  => 0,    # To auto-generate README from POD markup
    },
);

$builder->create_build_script;

exit;

sub pod2markdown {
    my ( $pod_file, $markdown_file ) = @_;
    open my $markdown_fh, ">", $markdown_file
        or die "Cannot open $markdown_file for writing: $!";
    my $parser = Pod::Markdown->new();
    $parser->output_fh($markdown_fh);
    $parser->parse_file($pod_file);
    close $markdown_fh;
}


lib/Log/History.pm
package Log::History;
use strict;
use warnings;
use Carp;
use Cwd ();
use POSIX;
use File::Copy 'move';
use File::Temp ();
use Scalar::Util 'looks_like_number';

our $VERSION = '0.1.0';

=head1 NAME

Log::History - Make self-documenting scripts that track their own execution history.

=head1 SYNOPSIS

Place the import statement for Log::History wherever you want the log to start.

    use Log::History;
    #2014-02-02 12:43:08 (00:00:48) in /path/to/workingdir: /path/to/script.pl --opt 1 --param 2
    ...

To limit the number of log entries, specify how many to keep.
The oldest entries are discarded.

    use Log::History '3';
    #2014-05-27 ...
    #2007-12-18 ...
    #1987-12-18 ...

=head1 DESCRIPTION

Log::History enables scripts to keep track of their own execution history.
Each log entry reports the date, start time, elapsed run time, working directory,
and a record of exactly how the script was called.

=head1 CAVEATS

Because Log::History modifies the script that imports it, problems can occur
if too many instances of the script are run at one time.
(Actually, the problem can arise if one instance ends at the exact time another is starting.)

If you don't have write access for the script, no log will be generated.

=head1 ACKNOWLEDGEMENTS

I was inspired to write this after recently re-reading Neil Bowers' post:
L<Identifying CPAN distributions you could help out with|http://blogs.perl.org/users/neilb/2012/12/modules-that-are-candidates-for-helping-out.html>.
In it, I found Tushar Murudkar's no-longer-maintained module L<Log::SelfHistory>.
I was intrigued since I had been working on L<Log::Reproducible|https://github.com/mfcovington/Log-Reproducible>.
I wanted to go in a different direction, so decided to start from scratch instead of trying to take over L<Log::SelfHistory>.

=head1 AUTHOR

Michael F. Covington <mfcovington@gmail.com>

=head1 SEE ALSO

L<Log::SelfHistory>, L<Log::Reproducible|https://github.com/mfcovington/Log-Reproducible>

=cut

sub _now {
    my %now;
    my @localtime = localtime;
    $now{'timestamp'} = strftime "%Y-%m-%d %H:%M:%S", @localtime;
    $now{'seconds'} = time();
    return \%now;
}

sub _elapsed {
    my ( $start_seconds, $finish_seconds ) = @_;

    my $secs = difftime $finish_seconds, $start_seconds;
    my $mins = int $secs / 60;
    $secs = $secs % 60;
    my $hours = int $mins / 60;
    $mins = $mins % 60;

    return join ":", map { sprintf "%02d", $_ } $hours, $mins, $secs;
}

my $start;
my $cwd;
my $script;
my @arguments;
my $package;
my $log_limit;
my $original_permissions;
my $was_imported = 0;

sub import {
    ( $package, $log_limit ) = @_;
    $was_imported = 1;

    @arguments = @ARGV;
    for (@arguments) {
        $_ = "'$_'" if /\s/;
    }
}

BEGIN {
    $start  = _now();
    $cwd    = getcwd();
    $script = $0;

    my $mode = ( stat($script) )[2];
    $original_permissions = sprintf "%04o", $mode & 07777;
}

END {
    return unless $was_imported;
    if ( !-w $script ) {
        warn
            "Warning: $package did not create a log, because you do not have write access for: $script\n";
        return;
    }

    my $finish = _now();
    my $elapsed = _elapsed( $$start{'seconds'}, $$finish{'seconds'} );

    my $log = "#$$start{'timestamp'} ($elapsed) in $cwd: $script";
    $log .= " @arguments" if @arguments;
    $log .= "\n";

    if ( !defined $log_limit ) {
        $log_limit = -1;
    }

    if ( !defined $log_limit || !looks_like_number $log_limit ) {
        $log_limit = -1;
my $warning = <<EOF;
Warning: A non-numeric log limit was passed to Log::History while running $script
         I hope you are a history buff, because your history is now... UNLIMITED!
EOF
        warn $warning;
    }

    # read script and insert item in log
    my @code;
    my $log_regex
        = qr/^#\d{4}-\d{2}-\d{2}\s\d{2}:\d{2}:\d{2}\s\(\d{2,}:\d{2}:\d{2}\)\sin/;
    open my $script_in_fh, "<", $script;
    flock( $script_in_fh, 2 ) or die $!;
    while ( my $line = <$script_in_fh> ) {
        push @code, $line;
        if ( $line =~ /^[^#]*use\s+$package/ ) {
            push @code, "\n" unless $line =~ /\n$/;
            push @code, $log;
            my $log_count = 1;
            while ( my $post_log_line = <$script_in_fh> ) {
                if ( $post_log_line =~ /$log_regex/ ) {
                    $log_count++;
                    next if $log_count > $log_limit && $log_limit != -1;
                }
                push @code, $post_log_line;
            }
        }
    }
    close $script_in_fh;

    # write script with newly logged item
    my ( $temp_fh, $temp_filename )
        = File::Temp::tempfile( DIR => $ENV{'HOME'} );
    print $temp_fh @code;
    close $temp_fh;
    move $temp_filename, $script;
    chmod oct($original_permissions), $script;
}

1;
